### PR TITLE
[ORCA] Stop checking data access indication

### PR DESCRIPTION
**Pull request by d**
The SQL 2003 standard specified 4 different levels of "SQL data access
indication":

1. NO SQL
2. CONTAINS SQL
3. READS SQL DATA
4. MODIFIES SQL DATA

Upstream PostgreSQL doesn't support any of the above. Greenplum has had
syntactic support for this as part of `CREATE FUNCTION` command, but so
far we only store that in the catalog (and possibly dump back out),
without taking it into consideration in our planning or execution
process. Except ORCA does, for a seemingly unnecessary purpose: ORCA
opts out of various optimizations when it sees a function that is not
"NO SQL" (that's a mouthful). The intent, it seems, is we really want to
execute such functions on the QD. But that seems questionable for two
fundamental reasons:

1. Forgoing the various optimizations (as evidenced by the test cases in
   this patch) does NOT ensure we execute these functions on the QD
   node.
1. "executing on QD" is only an approximate intent, the real desired
   outcome is that we want to enable a *dispatch* from within functions
   that contain SQL. But we don't have any control over that, either:
   often an active dispatch is already in-flight and the plan is
   effectively non-executable.
1. There is a small niche that *does* match the "execute on QD" intent,
   namely, querying a "master-only" table (such tables also exist on the
   segments, but they contain no data) in a SQL function. That's where
   the `EXECUTE ON COORDINATOR` syntax is the best signal.

TL;DR: we are effectively pessimizing for any function that possibly
contains SQL, without ensuring that we can achieve the intended goal
(dispatch nested queries from QD). We should rethink whether the SQL
data access indication is even the right signal for us to achieve such
goals (I don't think so).

This patch still keeps the data access attribute in metadata (DXL), but
removes it from the rest of ORCA, including conditions that used to
consider them alongside volatile. As a result, we can expect various
more optimal plans in different situations. See the patch for details.

### CODE REVIEW

**Comment by d** [comment]
When backported, this patch has the hope of being the less invasive alternative to what #11435 tries to achieve on 6X_STABLE

**Comment by l-wang** [comment]
Thank you @d and @khannaekta for the nice commit message and tests, I found them very helpful. I'm starting to review this.

**Comment by l-wang** [comment]
If nobody uses the data access property anymore in GPDB and never does Postgres, should we add a deprecate message in the parser?

**Comment by ashwinstar** [comment]
@maksm90 fyi.. you might be interested in this PR

**Comment by d** [comment]
> should we at least add a deprecate message in the parser?

I personally believe so, but the decision is a bit out-of-scope for this patch. @ashwinstar might be a better person to track that

File: src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp [hunk 1]
```diff
@@ -543,8 +543,7 @@ CTranslatorDXLToExpr::PexprLogicalTVF(const CDXLNode *dxlnode)
 
 	const IMDFunction *pmdfunc = m_pmda->RetrieveFunc(mdid_func);
 
-	if (IMDFunction::EfsVolatile == pmdfunc->GetFuncStability() ||
-		IMDFunction::EfdaNoSQL != pmdfunc->GetFuncDataAccess())
+	if (IMDFunction::EfsVolatile == pmdfunc->GetFuncStability())
 	{
 		COptCtxt::PoctxtFromTLS()->SetHasVolatileOrSQLFunc();
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp:548]: Rename this

**Review by chrishajas — APPROVED** [review]
Thanks for the great explanation! Overall LGTM. Are there other possible user effects from this change beside faster running queries (eg: is it possible that a function now doesn't execute since it previously executed on dispatch and now is running on segment?)

File: src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp (hunk 1 above)
Comment (d) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp:548]: shoot, this was totes on my mind when I was working on it. How about calling it `SetHasVolatile` and `GetHasVolatile`? Do I put a `Get` prefix there??

**Review by l-wang — APPROVED** [review]
I have a few questions and asked @d offline, posting them here for other reviews and my future self:

> Q1. Before this patch, functions that "CONTAINS/READS/MODIFIES SQL" are only allowed to execute on QD, after this patch, such functions will be allowed to execute on a singleton segment. If the intention was to not execute them on QD, would lifting this restriction from the optimizer cause a problem?

No. Say I have a function:
```sql
CREATE OR REPLACE FUNCTION foo() RETURNS integer AS $$
select count(*)::int from bar;
$$ LANGUAGE SQL;
```
where foo() contains sql and is not volatile. `select foo() from bar;` would fail anyways w/ or w/o the patch, because we cannot dispatch the execution of foo() when we are already on a segment.

Q2: It seems previously we always consider `volatile` and not NO SQL together, so whatever problem with function data access we are trying to solve here, would the same apply to function stability?

File: src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp (hunk 1 above)
Comment (chrishajas) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp:548]: Yep, those sound reasonable! I've seen patterns in the codebase with and without `Get`, so I don't have a strong opinion on either.

File: src/backend/gporca/libgpopt/src/operators/CLogicalRowTrigger.cpp [hunk 2]
```diff
@@ -33,9 +33,7 @@ CLogicalRowTrigger::CLogicalRowTrigger(CMemoryPool *mp)
 	  m_rel_mdid(nullptr),
 	  m_type(0),
 	  m_pdrgpcrOld(nullptr),
-	  m_pdrgpcrNew(nullptr),
-	  m_efs(IMDFunction::EfsImmutable),
```
[the remark this answers is not in the document]
Comment (d) [reply src/backend/gporca/libgpopt/src/operators/CLogicalRowTrigger.cpp:37]: It's actually intentional

**Comment by d** [comment]
@l-wang @chrishajas wanna look at the two new fixup commits I pushed? I plan to keep the backfill MDP commit separate and fixup all other commits into one when merging

**Review by chrishajas — APPROVED** [review]
@d Looks good!

File: src/backend/gporca/libgpopt/include/gpopt/base/COptCtxt.h [hunk 3]
```diff
@@ -223,7 +223,7 @@ class COptCtxt : public CTaskLocalStorageObject
 		//      In such cases, it is safe to *always* enforce gather motion on master as there is no way to determine
```
Comment (l-wang) [inline src/backend/gporca/libgpopt/include/gpopt/base/COptCtxt.h:223]: Remove the comment above this line.

File: src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalRowTrigger.h [hunk 4]
```diff
@@ -41,7 +41,7 @@ class CLogicalRowTrigger : public CLogical
 	CColRefArray *m_pdrgpcrNew;
 
 	// stability
-	IMDFunction::EFuncStbl m_efs;
+	IMDFunction::EFuncStbl m_efs{IMDFunction::EfsImmutable};
```
Comment (l-wang) [inline src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalRowTrigger.h:44]: Thank you for making the default explicit.

**Review by l-wang — APPROVED** [review]
New commits look good to me. Just added a minor nitpick.
